### PR TITLE
Add Friendship model, associations, and tests

### PR DESCRIPTION
**Pull request by pointerish**
This PR implements the following:

- Adds `Friendship` model
- Adds associations to create friendships
- Adds `RSpec` tests for the model

### CODE REVIEW

**Review by arslanbisharat — CHANGES_REQUESTED** [review]
# Request changes 😳

I am very happy to see this project and I can see you worked hard on this project. There are only a few issues that we need to address and I am sure that you will end up with a perfect project

_It's okay. We are here to learn. Don't get discouraged. Work even harder. You are almost there.😄_

This is a great opportunity to learn. Microverse is here to help you. 

**Please have a look at the suggestions to get more understanding.**

- [x] Please update your Readme and make it more descriptive, 

- [x] There should be NO application logic in the views nor helpers (no Post.new, Comment.create, current_user.friends.find_by(...)...). Everything connected with data processing should be encapsulated in models and used in controllers.

**Milestone 5: Friendships v1**

- [x] User model `has_many :inverted_friendships, foreign_key: "friend_id"`


**Please make the changes and request a re-review for the given project**

If you get stuck somewhere, don't hesitate to reach me out or schedule s 1:1 meeting with a Full-time TSE. 

Stay safe and give me a smile 😇

### Happy Coding
---------------------------------------------------------------------------------------------------------------------------------
**Regards,
Arslan Bisharat
slack: arslanbisharat**

**Review by umairarshadbutt — CHANGES_REQUESTED** [review]
# Status: REQUIRED CHANGES :warning: :construction: 

Hi @pointerish 
Good job with the changes! :tada: We are almost there.
We, however, do have a few changes to carry out before we proceed (see review comment).

##

Kindly work on the suggested changes  👇.
- [ ] Please remove the logic from views.
- [ ] Rspec test cases are not workinig
- [x] MIT license giving the 404 error(page not found), please fix this issue. 

##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

**Comment by pointerish** [comment]
Hi @umairarshadbutt, I fixed the MIT license issue. 

Could you please elaborate on the rest? When I run `rspec spec` the tests run fine and succeed. Also, the only logic on the views comes from `application.html.erb` and it was already there to begin with.

**Comment by pointerish** [comment]
I resubmitted since I received no response. Thanks in advance!

**Review by rindrajosia — CHANGES_REQUESTED** [review]
**Changes required**
Hello!!
You did a very good job.
However, to move forward you have to make some changes:
REQUIRED:
- [x] In your` friendships` table, you need also to define `friendee_id` as foreign key:


- [x] **Friendship model** _belongs_to :user_ ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friend.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- [x] **Friendship model** `belongs_to :friend, class_name: "User"` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friend.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

As you can see in this screenshot, your `Friendship table` should have 2 foreign keys, so make sure that your `Friendship model` corresponds with that.

- [x] User can invite friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/send_invitation.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- [ ] User can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- [ ] User has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- [ ] User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- [ ] There is a button to invite user to friendship on users list page and on single user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))



You should `add a button to invite user to friendship` on these pages.

- [ ] I should be able to see **only** "Posts" **written by me** and **my friends**.([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

As you can see in this screenshot, Josia is not yet Marina's friend however, Marina can see Josia's Post.

Please, make the suggested changes and submit a new code review request
Thanks
Rindra Josia

File: app/models/friendship.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  include ActiveModel::Validations
+
+  validates_with SelfFriendshipValidator
+
+  validates :status, presence: true
+  validates :user_id, presence: true
+  validates :friendee_id, presence: true
+
+  belongs_to :user, foreign_key: "user_id", class_name: "User"
+  belongs_to :user, foreign_key: "friendee_id", class_name: "User"
```
Comment (blackpintz) [inline app/models/friendship.rb:17]: Please fix the above to this:
```
  belongs_to :user
  belongs_to :friend, class_name: "User"
```

File: app/models/friendship.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+  # def confirm_friend
+  #   create(friendee_id: current_user.id, user_id: id, status: 'confirmed')
+  # end
```
Comment (blackpintz) [inline app/models/friendship.rb:21]: Remove this commented code.

File: app/models/friendship.rb (hunk 1 above)
Comment (blackpintz) [inline app/models/friendship.rb:17]: Write tests for the above associations in `friendship_spec.rb`.

**Review by blackpintz — CHANGES_REQUESTED** [review]
# Status: Changes Required 🛑 

### Hello team 👋,

**Great job implementing the requested changes 👏. However, there are still a few fixes needed.**

1). **I create a post but it doesn't show on the timeline.**
2). **A logged-in user should also be listed in the users' list page. In the example below, Mercy's name should also be listed next to Jen's name.**
users

3). **After a friend request button is sent, the button should switch to a text informing the user that the friend request has been sent. See screenshot below.**

 before a request is sent  | After a request is sent
 ---------------------------| ------------------------
before_request | after_request

4). **Say, Mercy sends Jen a friend request. When Jen logs in, she should not see a friend request button next to Mercy's name and her profile.**

5). **When Jen accepts Mercy's friend request, she should not see an add friend button next to Mercy's name. Instead, she should see a text like, "You are friends."**

6). **Please add all your feature tests in `spec/features` directory and model tests in `spec/models` directory.**

7). **Add tests for your models. You only wrote tests for the Friendship model.**

8). **You have a failing test as shown in the screenshot below..**
rspec

9). **Check more in the comments below.**

## Happy coding 🥂.

**You can reach me on slack (Rose Wanjohi).**

File: app/models/friendship.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+class Friendship < ApplicationRecord
+  include ActiveModel::Validations
+
+  validates_with SelfFriendshipValidator
+
+  validates :status, presence: true
+  validates :user_id, presence: true
+  validates :friendee_id, presence: true
+
+  belongs_to :user, foreign_key: 'user_id', class_name: 'User'
```
Comment (SmithVyne) [inline app/models/friendship.rb:20]: `belongs_to :user
 belongs_to :friend, class_name: 'User'`

**Review by SmithVyne — CHANGES_REQUESTED** [review]
Just a few more minor changes👋 👋  
Hello, congratulations on making all the previous changes required by the TSEs, but you need to update your friendship model.

File: app/models/friendship.rb (hunk 3 above)
Comment (SmithVyne) [inline app/models/friendship.rb:20]: It would be most correct to just say:
`
belongs_to :user
 belongs_to :friend, class_name: 'User'
`

**Review by oloomoses — CHANGES_REQUESTED** [review]
# Status: Changes Required
Hello team, You really worked hard on this project and after so much feedback, I can assure you that you are almost done. Just some minor issues that I am sure will not take you long.
- [x] Check that your linter is configured correctly and there are no rubocop errors. you can check to see if the GitHub linter is enabled
- [ ] Some RSpec tests are still failing. Kindly fix them.
Fix the above issues before submitting another review.

# Slack [Oloo Moses

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Josias Alvarado Gutierrez** 
Good job with your improvements for this project :muscle:  :muscle:
It meets most of the requirements :clap: :clap: :clap:  
There are just a few details to improve before you can go ahead:

- [ ] Fix rubocop offenses
- [x] Don't use multiple expect on the same scenario
- [ ] Don't overuse rubocop disable
- [ ] After I accept/decline a request friendship it show me the option to add myself (current user) as a friend, that shouldn't happen: 

- [ ] Add a screenshot in your README file
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Review by kobiyoyo — APPROVED** [review]
# Project Approved :thumbsup:
## Hello Team  ,  you have done a brilliant job with this project. :clap:
#### You can merge !!!

**Pull request merged**